# Module 2: From `datascience` to `pandas`

## Why `pandas`?

Like the `datascience` package you learned and used in Data 8, `pandas` is a Python library used for data manipulation and analysis.
However, `datascience` was developed as a pedagogical tool for Data 8, intended to help students become familiarized with python syntax as well as syntax associated with tabular data analysis. 
We decided to teach `datascience` first since its syntax is more intuitive and easier to use for students without much programming experience.
On the other hand, `pandas` is an industrial strength package that is used in most data analysis projects in the real world. 
Learning how to use pandas would also make your projects easier to understand for other data scientists and extend the scope of influence your projects may have.


Now that you've completed Data 8, it's a good time to translate the functions you've learned to `pandas`. 
Throughout this notebook, we will go over `pandas` by showing similar functions in both the `datascience` syntax and `pandas` syntax, as well as introduce you to some of the other functionalities that `pandas` provides.

We will first import the packages we need. The following 3 lines imports `pandas`, `datascience`, and `numpy` respectively. 
Note that we imported `pandas` as `pd`, which means that when we call functions in `pandas`, we will always reference them with `pd` first.
In Data 8, you saw something similar when we imported `numpy` as `np`: functions like taking `np.mean` or `np.random.choice` are all from the `numpy` package.


In [1]:
import pandas as pd # This is how we import pandas into the environment. Typically, we use pd to refer to all pandas modules for short. 
from datascience import *
import numpy as np

## Creating a Table

### Reading in a Dataset

Most of the time, the data we want to analyze will be in a separate file, typically as a `.csv` file. 
In this case, we want to read the files in and convert them into a tabular format.

Using the `datascience` package, we will want to read in the file as a table. 
To do this, we use the function `Table.read_table(file_path)`. 
In the example below, the `baby.csv` file is in the same folder as this notebook, so the relative file path from this notebook to the csv file is just `baby.csv`.

In [2]:
# datascience
baby_tbl = Table.read_table("baby.csv")
baby_tbl.show(5)

The syntax for reading in csv files in `pandas` is almost identical. 
`pandas` has a specific function to read in csv files called `pd.read_csv(file_path)`, with the same relative file path as its argument. 
the `dataframe.head()` function will display the first 5 rows of the data frame by default. 
If you want to specify the number of rows displayed, you can use`dataframe.head(num_rows)`.
Similarly, if you want to see the last few rows of the data frame, you can use `dataframe.tail(num_rows)`. Try it out for yourself!

In [3]:
# pandas
baby_df = pd.read_csv("baby.csv")
baby_df.head(5)

Birth.Weight  Gestational.Days  Maternal.Age  Maternal.Height  \
0           120               284            27               62   
1           113               282            33               64   
2           128               279            28               64   
3           108               282            23               67   
4           136               286            25               62   

   Maternal.Pregnancy.Weight  Maternal.Smoker  
0                        100            False  
1                        135            False  
2                        115             True  
3                        125             True  
4                         93            False

### Creating a Table from Scratch
Often times, data in tabular format is preferred for analysis. 
But what if you already have the data, just not in tabular format? 
For example, if your data is scattered in arrays and lists, you can use that to build your own table from scratch.

Using the `datascience` package, we can first create a blank table with a call to Table(), 
then add in the arrays as columns to this blank table as shown below.

In [4]:
# datascience
flowers_tbl = Table().with_columns(
    'Number of petals', make_array(8, 34, 5),
    'Name', make_array('lotus', 'sunflower', 'rose'),
    'Color', make_array('pink', 'yellow', 'red')
) 
flowers_tbl

Number of petals | Name      | Color
8                | lotus     | pink
34               | sunflower | yellow
5                | rose      | red

There are multiple ways to do this in `pandas`. 
One way is intuitively very similar to how we did it with `datascience`. 

Here we use a dictionary object to represent the data -- don't worry if you're not familiar with them. 
We pass into the `data` argument `{"colname1": column1, "colname2": column2, ...}`.
Notably, we build the `DataFrame` by the columns, using each list as a column and associating each column with its appropriate name. 

In [5]:
# pandas - method 1
flowers_df = pd.DataFrame(data = {'Number of petals': [8, 34, 5], 
                         'Name': ['lotus', 'sunflower', 'rose'],
                         'Color': ['pink', 'yellow', 'red']})
flowers_df

Number of petals       Name   Color
0                 8      lotus    pink
1                34  sunflower  yellow
2                 5       rose     red

Another way to build the same table from scratch is by building it with rows. 
With this method, the data should be a list of lists, where each inner list contains the entries of one row. 
You might also notice that there is now a second argument, `columns`. 
Since we are passing in rows, we do not have the column names inside the `data` argument. 
This is why we use the `columns` argument to specify column names.

In [6]:
# pandas - method 2
flowers_df = pd.DataFrame(data = [[8, 'lotus', 'pink'], 
                                  [34, 'sunflower', 'yellow'],
                                  [5, 'rose', 'red']], 
                         columns = ['Number of petals', 'Name', 'Color'])
flowers_df

Number of petals       Name   Color
0                 8      lotus    pink
1                34  sunflower  yellow
2                 5       rose     red

## `pandas` datatypes

### Series vs. Arrays


One of the primary data types we saw when analyzing tabular data with the `datascience` package is the array.
In `datascience`, the columns of the tables consisted of arrays. In `pandas`, there is a very similar, but slightly different data type called a `Series`.


 You can access the values of a column in a Table using the `tbl.column(column_name)` function as follows. 
 When using `tbl.column`, the values of the selected column will be returned as an array.

In [7]:
## datascience
baby_tbl.column("Birth.Weight")

array([120, 113, 128, ..., 130, 125, 117])

Similarly in `pandas`, you can access the values of a particular column by using `dataframe[column_name]`. 
The data frame object is introduced in the next section, for now you can just understand it as the `pandas` equivalent to a table. 
`dataframe[column_name]` will return a `Series` instead of an array. 


In [8]:
## pandas
baby_df["Birth.Weight"]

0       120
1       113
2       128
3       108
4       136
5       138
6       132
7       120
8       143
9       140
10      144
11      141
12      110
13      114
14      115
15       92
16      115
17      144
18      119
19      105
20      115
21      137
22      122
23      131
24      103
25      146
26      114
27      125
28      114
29      122
       ... 
1144    114
1145    127
1146     87
1147    141
1148    144
1149    116
1150     75
1151    138
1152     99
1153    118
1154     97
1155    146
1156     81
1157    110
1158    135
1159    114
1160    124
1161    115
1162    143
1163    113
1164    109
1165    103
1166    118
1167    127
1168    132
1169    113
1170    128
1171    130
1172    125
1173    117
Name: Birth.Weight, Length: 1174, dtype: int64

A `Series` object is basically an array with indices for each data point. 
In the above example, the first element in the `Birth.Weight` column is the integer 120. The corresponding index is 0.

If we want just the data as an array without the index, we can use the `Series.values` attribute.

In [9]:
baby_df["Birth.Weight"].values

array([120, 113, 128, ..., 130, 125, 117], dtype=int64)

### DataFrames vs. Tables

The following is our standard `datascience` Table. It is basically a collection of arrays, with column names.

In [10]:
# datascience
baby_tbl.head()

AttributeError: 'Table' object has no attribute 'head'

A `pandas`  `DataFrame` can be thought of as a collection of Series, all of which have the same index. 
The resulting `DataFrame` consists of columns where each column is a `Series` and each row has a unique index.

In [ ]:
# pandas
baby_df.head()

The number of rows in the `Table` can be found as such:

In [ ]:
# datascience
baby_tbl.num_rows

Similarly, for the number of columns:

In [ ]:
# datascience
baby_tbl.num_columns

The number of rows and columns in a `DataFrame` can be accessed together using the  `.shape` attribute. 
Notice that the index is not counted as a column.

In [ ]:
# pandas
baby_df.shape

To get just the number of rows, we want the 0th element.

In [ ]:
# pandas
baby_df.shape[0]

For just the number of columns, we want the 1st element.

In [ ]:
# pandas
baby_df.shape[1]

### Indices

The row labels of a `DataFrame` are collectively called the index. 
It helps to identify each row. 
By default, the index values are the row numbers, with the first row having index 0.

In [ ]:
# pandas
baby_df.head()

We can access the index of a `DataFrame` by calling `DataFrame.index`.

In [ ]:
baby_df.index

That doesn't seem too meaningful. 
We can access the values of the index using `.values`.

In [ ]:
# pandas
baby_df.index.values

In addition, we can set the index to whatever we want it to be. 
So, instead of index going from 0 to 1173, we can change it to go from 1 to 1174.

In [ ]:
# pandas
baby_df.set_index(np.arange(1, 1175))

Let's look at another example.

In [ ]:
flowers_df

The labels in an index of a `DataFrame` do not have to be intergers; they can also be strings. 
We can also use one of the data columns to be the index itself. 

Here is an example in which we are setting the index to be the `Name` column.

In [ ]:
# pandas
flowers_df = flowers_df.set_index('Name')
flowers_df

## Subsetting Data

### Selecting Columns

Sometimes the entire dataset contains too many columns, and we are only interested in some of the columns. 
In these situations, we would want to be able to select and display a subset of the columns from the original table. 
We dicuss some of these methods below. 

In the `datascience` package, the `select` function is used. 
If we want the returned columns to be a table as well, we use the syntax `tbl.select(col_name1, col_name2, ...)`

In [ ]:
# datascience
# Selects the columns "Number of petals" and "Color".
flowers_tbl.select("Number of petals", "Name")

In `pandas`, there are many ways to achieve the same result. For one, we can use the function `loc`, shown below. 

The first argument of `loc` is which rows we want to select, and since we want all of the rows, just a colon ":" would indicate all rows. 
The second argument selects the columns we want. 

If we want more than one column, we need to pass in the column names as a list for the `loc` to return a dataframe object. 

In [ ]:
# pandas
# Selects all rows and the columns "Number of petals" and "Color".
flowers_df.loc[:,["Number of petals", "Color"]]

If you pass in a single column name as a list, it will still return a dataframe object with one column.

In [ ]:
# pandas
# Selects all rows but only the column "Number of petals". Returns a DataFrame object.
flowers_df.loc[:,["Number of petals"]]

But if you pass in the column name as a string, pandas will recognize that you only have one column, and return a `Series` instead.

In [ ]:
# pandas
# Selects all rows but only the column "Number of petals". Returns a series object.
flowers_df.loc[:,"Number of petals"]

Another way to subset data in `pandas` is to use `iloc`. 

Unlike `loc`, which accepts column names as arguments, `iloc` only accepts numerical indices as its arguments. 
The order of arguments remain the same, with the rows being the first argument and the columns being the second argument. 
Here the `0:1` denotes a range and means that we want all columns indexed 0 through 1. 
In Python, ranges are generally left inclusive and right exclusive (so that only column 0 is selected here).

In [ ]:
# pandas
# Selects all rows and the 0th-1st column (not inclusive of the 1st column)
flowers_df.iloc[:,0:1]

As a side note, we can also subset tables' rows by selecting row indices. 
Since we've set the index of the flowers table to be the name of the flower, we can directly pass in the row indices as a list as the first argument. 
Note that here, `loc` is actually left and right inclusive.

In [ ]:
# pandas
# Selects the rows with index values "sunflower" and "lotus", and all columns
flowers_df.loc[["sunflower", "lotus"], :]

In [ ]:
# pandas
# Selects the rows with index values from 0 to 3 (inclusive of 0 and 3), and all columns
baby_df.loc[0:3,:]

`loc` and `iloc` are very powerful functions in `pandas`. 
Here are 2 more examples on the `baby` table, let's see what they do:

In [ ]:
# pandas
# Selects the 0th and 2nd column, and all rows of the table
baby_df.iloc[:, [0,2]]

If we want to only select a subset of columns from the table, there exists a special case short-cut where we drop the `.loc` and `:` entirely:

In [ ]:
# pandas
# Selects the columns "Birth.Weight" and "Maternal.Age" with all rows
baby_df[["Birth.Weight", "Maternal.Age"]]

### Getting a Value

What if you want to single out one entry of your entire table? 
This often occurs when we want to max or min value after sorting the table, for example:
- *What is the name of the flower with the most number of petals?*
- *How heavy was the baby that went through the longest gestational days?*

When we want a specific entry using the `datascience` package, we need to first use `tbl.column` to fetch an array out of a table, then use `.item` to retrieve the element's value. 
In the code below, we get the birth weight of the first baby recorded in this dataset. 

In [ ]:
# datascience
# Get the first item from the "Birth.Weight" column
baby_tbl.column("Birth.Weight").item(0)

In `pandas`, the syntax for getting a single element is a lot less verbose. 
Remember `loc` and `iloc`? Since these functions have the ability to subset rows and columns at the same time, we are going to use that functionality here. 

We pass in 0 as the row selector, since we only want the first entry of the "Birth.Weight" column (the entry at the 0th index).

In [ ]:
# pandas
# Get the value at row with index 0 and column with label "Birth.Weight".
baby_df.loc[0,"Birth.Weight"]

Similarly with iloc, we are just passing in 0 as the first and second argument since we want the entry located at the first row and first column, which are both indexed at 0.

In [ ]:
# pandas
# Get the value at the 0th row and 0th column.
baby_df.iloc[0,0]

In [ ]:
# pandas
# Get the rows with indices 0 to 5 (inclusive) of the "Birth.Weight" column
baby_df.loc[0:5, "Birth.Weight"]

In [ ]:
# pandas
# Select the first five columns of the first two rows
baby_df.iloc[0:2, 0:5]

# Methods

## Filtering and Boolean Indexing

With the `datascience` package, we can filter a table by only returning rows that satisfy a specific condition. 

In [ ]:
# datascience
# Returns all of the rows where "Birth.Weight" is greater than 120
baby_tbl.where('Birth.Weight', are.above(120))

Equivalently, we can do this in `pandas` by "boolean indexing". 
The expression below returns a boolean series where an entry is `True` if it satisfies the condition and `False` if it doesn't.

In [ ]:
# pandas
baby_df['Birth.Weight'] > 120

If we want to filter our `pandas` dataframe for all rows that satisfies Birth.Weight > 120, we can pass the boolean series into the row argument of `.loc`. 
The idea is that we only want the rows where the "boolean index" is `True`. 

In [ ]:
# pandas
# Select all rows that are True in the boolean series baby_df['Birth.Weight'] > 120.
baby_df.loc[baby_df['Birth.Weight'] > 120, :]

Notably, `.loc` returns all columns by default so we can omit the column argument and get the same result.

In [ ]:
# pandas
# Select all rows that are True in the boolean series baby_df['Birth.Weight'] > 120.
baby_df.loc[baby_df['Birth.Weight'] > 120]

Boolean indexing is a very popular way to conduct filtering in `pandas`. As such, there exists another special case short-ahnd where we don't need the `.loc` or the `:`.

In [ ]:
# pandas
# Select all rows that are True in the boolean series baby_df['Birth.Weight'] > 120.
baby_df[baby_df['Birth.Weight'] > 120]

In general, a filtering expression of the form `tbl.where(column, predicate)` in the `datascience` library takes the form `df.loc[criterion]` in `pandas`. 
Here are a few more examples:

In [ ]:
# datascience
# Return all rows where Maternal.Height is greater than or equal to 63.
baby_tbl.where('Maternal.Height', are.above_or_equal_to(63))

In [ ]:
# pandas
# Return all rows where Maternal.Height is greater than or equal to 63.
baby_df[baby_df['Maternal.Height'] >= 63]

In [ ]:
# datascience
# Return all rows where Maternal.Smoker is True.
baby_tbl.where('Maternal.Smoker', are.equal_to(True))

In [ ]:
# pandas
# Return all rows where Maternal.Smoker is True.
baby_df.loc[baby_df['Maternal.Smoker'] == True]

## Filtering on Multiple Conditions

We can also filter on multiple conditions. 
If we want records (rows) where all of the conditions are true, we separate our criterion by the `&` symbol, where `&` represents *and*.

`df.loc[(boolean series 1) & (boolean series 2) & (boolean series 2)]`

If we just want one of the conditions to be true, we separate our criterion by `|` symbols, where `|` represents *or*.

`df.loc[(boolean series 1) | (boolean series 2) | (boolean series 2)]`




In [ ]:
# datascience
# Return all rows where Gestational.Days is between 270 and 280.
baby_tbl.where('Gestational.Days', are.between(270, 280))

In [ ]:
# pandas
# Select all rows where Gestational.Days are above or equal to 270, but less than 280.
baby_df.loc[(baby_df['Gestational.Days'] >= 270) & (baby_df['Gestational.Days'] < 280)]